sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -327,121 +329,335 @@ bool WinHttpTransport::VerifyCertificatesInChain(
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_DATA_AVAILABLE);
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_READ_COMPLETE);
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_WRITE_COMPLETE);
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_REQUEST_ERROR);
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_SENDREQUEST_COMPLETE);
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_GETPROXYFORURL_COMPLETE);
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_CLOSE_COMPLETE);
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_SHUTDOWN_COMPLETE);
+  // For <reasons> this is not defined on the Win2022 Azure DevOps image, so manually expand it.
+  //  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_GETPROXYSETTINGS_COMPLETE);
+  if (internetStatus & 0x08000000)
   {
-    return;
+    rv += std::string("WINHTTP_CALLBACK_STATUS_GETPROXYSETTINGS_COMPLETE") + " ";
   }
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_SETTINGS_WRITE_COMPLETE);
+  APPEND_ENUM_STRING(WINHTTP_CALLBACK_STATUS_SETTINGS_READ_COMPLETE);
+  return rv;
+}
+#undef APPEND_ENUM_STRING
 
-  try
+std::string GetErrorMessage(DWORD error)
+{
+  std::string errorMessage = " Error Code: " + std::to_string(error);
+
+  char* errorMsg = nullptr;
+  if (FormatMessage(
+          FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_FROM_HMODULE | FORMAT_MESSAGE_ALLOCATE_BUFFER,
+          GetModuleHandle("winhttp.dll"),
+          error,
+          MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+          reinterpret_cast<LPSTR>(&errorMsg),
+          0,
+          nullptr)
+      != 0)
+  {
+    // Use a unique_ptr to manage the lifetime of errorMsg.
+    std::unique_ptr<char, decltype(&LocalFree)> errorString(errorMsg, &LocalFree);
+    errorMsg = nullptr;
+
+    errorMessage += ": ";
+    errorMessage += errorString.get();
+  }
+  errorMessage += '.';
+  return errorMessage;
+}
+} // namespace
+
+namespace Azure { namespace Core { namespace Http { namespace _detail {
+
+  /**
+   * Wait for an action to complete.
+   *
+   */
+  bool WinHttpAction::WaitForAction(
+      std::function<void()> initiateAction,
+      DWORD expectedCallbackStatus,
+      Azure::Core::Context const& context,
+      Azure::DateTime::duration const& pollInterval)
+  {
+    context.ThrowIfCancelled();
+    ResetEvent(m_actionCompleteEvent.get());
+    m_expectedStatus = expectedCallbackStatus;
+    m_stowedError = 0;
+    m_stowedErrorInformation = 0;
+    m_bytesAvailable = 0;
+
+    // Call the provided callback to start the WinHTTP action.
+    initiateAction();
+
+    DWORD waitResult;
+    do
+    {
+      waitResult = WaitForSingleObject(
+          m_actionCompleteEvent.get(),
+          static_cast<DWORD>(
+              std::chrono::duration_cast<std::chrono::milliseconds>(pollInterval).count()));
+      if (waitResult == WAIT_TIMEOUT)
+      {
+        context.ThrowIfCancelled();
+      }
+      else if (waitResult != WAIT_OBJECT_0)
+      {
+        return false;
+      }
+    } while (waitResult != WAIT_OBJECT_0);
```

Comment:
Interesting, how come the "else-if return" and "while-loop" conditions, identical?

Just a nit: The way the code is written here, I have to really think about how the control flow changes for different `waitResult` values. Consider writing differently.

Reply:
Also, how is the code coverage for `win_http_transport.cpp`?

Our code coverage CI leg only runs on linux/curl, so we don't see the coverage number for Windows-specific transport adapter files:
https://dev.azure.com/azure-sdk/public/_build/results?buildId=1934177&view=codecoverage-tab

https://github.com/Azure/azure-sdk-for-cpp/blob/53c9da155244b72ac90ad27f515d4ea1e7505a01/eng/pipelines/templates/stages/platform-matrix.json#L143-L149

https://github.com/Azure/azure-sdk-for-cpp/blob/53c9da155244b72ac90ad27f515d4ea1e7505a01/eng/pipelines/templates/stages/platform-matrix-live.json#L37-L43

Reply:
The `while` loop terminates when the event is set to the signalled state, otherwise it continues to run.
There are three possible return values from WaitForSingleObject:

1.  WAIT_OBJECT_0 - the event was set to the signalled state.
1. WAIT_TIMEOUT - the wait timed out
1. Anything else - There was an error.

The if/else clauses check for those three cases - if WaitForSingleObject returns WAIT_TIMEOUT, we check for cancellation and loop back to wait some more. If WaitForSingleObject returns WAIT_OBJECT_0, we return, if WaitForSingleObject returns anything else, we return false, which lets the caller know to throw an exception.

And I have literally no idea what the code coverage numbers are on Windows since we don't have the ability to measure it.